{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TreeDLib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from feature_templates import *\n",
    "from basic_features import *\n",
    "from treedlib_util import PTSVParser\n",
    "import lxml.etree as et\n",
    "from treedlib_structs import corenlp_to_xmltree, XMLTree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We define three classes of operators:\n",
    "* _NodeSets:_ $S : 2^T \\mapsto 2^T$\n",
    "* _Indicators:_ $I : 2^T \\mapsto \\{0,1\\}^F$\n",
    "* _Combinators:_ $C : \\{0,1\\}^F \\times \\{0,1\\}^F \\mapsto \\{0,1\\}^F$\n",
    "\n",
    "where $T$ is a given input tree, and $F$ is the dimension of the feature space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Basic IO\n",
    "parser = PTSVParser([\n",
    "        ('doc_id', 'text'),\n",
    "        ('sent_id', 'int'),\n",
    "        ('text', 'text'),\n",
    "        ('words', 'text[]'),\n",
    "        ('lemmas', 'text[]'),\n",
    "        ('poses', 'text[]'),\n",
    "        ('ners', 'text[]'),\n",
    "        ('char_idxs', 'int[]'),\n",
    "        ('dep_paths', 'text[]'),\n",
    "        ('dep_parents', 'int[]')\n",
    "    ])\n",
    "rows = map(parser.parse_line, open('test/test1.parsed.tsv', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Row(dep_paths=['nn', 'nsubj', 'cop', 'det', 'amod', 'amod', 'amod', 'nn', None, 'vmod', None, 'agent', None, 'det', 'nn', 'nn', 'prep_of', None], char_idxs=[0, 11, 20, 23, 25, 30, 38, 46, 63, 71, 78, 81, 91, 94, 98, 103, 110, 114], text=Hyper-IgM1 syndrome is a rare genetic primary immunodeficiency disease caused by mutations of the CD40 ligand gene., ners=['O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O'], sent_id=1, words=['Hyper-IgM1', 'syndrome', 'is', 'a', 'rare', 'genetic', 'primary', 'immunodeficiency', 'disease', 'caused', 'by', 'mutations', 'of', 'the', 'CD40', 'ligand', 'gene', '.'], lemmas=['hyper-igm1', 'syndrome', 'be', 'a', 'rare', 'genetic', 'primary', 'immunodeficiency', 'disease', 'cause', 'by', 'mutation', 'of', 'the', 'cd40', 'ligand', 'gene', '.'], dep_parents=[2, 9, 9, 9, 9, 9, 9, 9, 0, 9, 0, 10, 0, 17, 17, 17, 12, 0], poses=['NN', 'NN', 'VBZ', 'DT', 'JJ', 'JJ', 'JJ', 'NN', 'NN', 'VBN', 'IN', 'NNS', 'IN', 'DT', 'NN', 'NN', 'NN', '.'], doc_id=0)>"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing XML speeds\n",
    "\n",
    "How does it compare between:\n",
    "* parse to XML via this python code, store as string, then parse from string at runtime\n",
    "* just parse to XML at runtime via this python code?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 8.92 ms, sys: 2.56 ms, total: 11.5 ms\n",
      "Wall time: 9.56 ms\n"
     ]
    }
   ],
   "source": [
    "# Map sentence to xmltree\n",
    "%time xts = map(corenlp_to_xmltree, rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 433 µs, sys: 326 µs, total: 759 µs\n",
      "Wall time: 1.21 ms\n"
     ]
    }
   ],
   "source": [
    "# Pre-process to xml string\n",
    "xmls = [xt.to_str() for xt in map(corenlp_to_xmltree, rows)]\n",
    "\n",
    "# Parse @ runtime using lxml\n",
    "%time roots = map(et.fromstring, xmls)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple demo: Generating DDLib features\n",
    "\n",
    "As a first simple demo let's generate the features generated by [ddlib](http://deepdive.stanford.edu/doc/basics/gen_feats.html).\n",
    "\n",
    "**_Note there are some noticeable differences stemming from using a dep tree representation at base; but a simple linear representation of the sentence could be used too, as in DDLib_**\n",
    "\n",
    "First, let's load a few sample sentences and convert one of them to XML format for testing; we'll also load the feature templates library, and also tag some candidates (crudely for now) to play around with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".node {\n",
       "  cursor: pointer;\n",
       "}\n",
       "\n",
       ".node circle {\n",
       "  fill: #fff;\n",
       "  stroke: steelblue;\n",
       "  stroke-width: 3px;\n",
       "}\n",
       "\n",
       ".node text {\n",
       "  font: 12px sans-serif;\n",
       "}\n",
       "\n",
       ".edge {\n",
       "  fill: none;\n",
       "  stroke: #ccc;\n",
       "  stroke-width: 2px;\n",
       "  cursor: pointer;\n",
       "}\n",
       "\n",
       ".highlight {\n",
       "  stroke: red;\n",
       "  stroke-width: 3px;\n",
       "}\n",
       "</style>\n",
       "\n",
       "<!--Provide the canvas id via python string formatting here--!>\n",
       "<div id=\"tree-chart-6391770092561218128\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "$.getScript(\"http://d3js.org/d3.v3.min.js\", function () {\n",
       "// See http://bl.ocks.org/d3noob/8375092\n",
       "// Two vars need to be provided via python string formatting: the tree, and the canvas id\n",
       "\n",
       "// JSON tree\n",
       "var root = {\"attrib\": {\"word\": \"disorder\", \"ner\": \"O\", \"pos\": \"NN\", \"char_idx\": \"80\", \"lemma\": \"disorder\", \"word_idx\": \"11\", \"dep_parent\": \"0\"}, \"children\": [{\"attrib\": {\"word\": \"disease\", \"dep_parent\": \"12\", \"pos\": \"NN\", \"char_idx\": \"37\", \"lemma\": \"disease\", \"word_idx\": \"4\", \"dep_path\": \"nsubj\", \"ner\": \"O\"}, \"children\": [{\"attrib\": {\"word\": \"Autosomal\", \"dep_parent\": \"5\", \"pos\": \"JJ\", \"char_idx\": \"0\", \"lemma\": \"autosomal\", \"word_idx\": \"0\", \"dep_path\": \"amod\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"dominant\", \"dep_parent\": \"5\", \"pos\": \"JJ\", \"char_idx\": \"10\", \"lemma\": \"dominant\", \"word_idx\": \"1\", \"dep_path\": \"amod\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"polycystic\", \"dep_parent\": \"5\", \"pos\": \"JJ\", \"char_idx\": \"19\", \"lemma\": \"polycystic\", \"word_idx\": \"2\", \"dep_path\": \"amod\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"kidney\", \"dep_parent\": \"5\", \"pos\": \"NN\", \"char_idx\": \"30\", \"lemma\": \"kidney\", \"word_idx\": \"3\", \"dep_path\": \"nn\", \"ner\": \"O\"}, \"children\": []}]}, {\"attrib\": {\"word\": \"is\", \"dep_parent\": \"12\", \"pos\": \"VBZ\", \"char_idx\": \"45\", \"lemma\": \"be\", \"word_idx\": \"5\", \"dep_path\": \"cop\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"the\", \"dep_parent\": \"12\", \"pos\": \"DT\", \"char_idx\": \"48\", \"lemma\": \"the\", \"word_idx\": \"6\", \"dep_path\": \"det\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"common\", \"dep_parent\": \"12\", \"pos\": \"JJ\", \"char_idx\": \"57\", \"lemma\": \"common\", \"word_idx\": \"8\", \"dep_path\": \"amod\", \"ner\": \"O\"}, \"children\": [{\"attrib\": {\"word\": \"most\", \"dep_parent\": \"9\", \"pos\": \"RBS\", \"char_idx\": \"52\", \"lemma\": \"most\", \"word_idx\": \"7\", \"dep_path\": \"advmod\", \"ner\": \"O\"}, \"children\": []}]}, {\"attrib\": {\"word\": \"human\", \"dep_parent\": \"12\", \"pos\": \"JJ\", \"char_idx\": \"64\", \"lemma\": \"human\", \"word_idx\": \"9\", \"dep_path\": \"amod\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"monogenic\", \"dep_parent\": \"12\", \"pos\": \"JJ\", \"char_idx\": \"70\", \"lemma\": \"monogenic\", \"word_idx\": \"10\", \"dep_path\": \"amod\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"caused\", \"dep_parent\": \"12\", \"pos\": \"VBN\", \"char_idx\": \"96\", \"lemma\": \"cause\", \"word_idx\": \"14\", \"dep_path\": \"conj_and\", \"ner\": \"O\"}, \"children\": [{\"attrib\": {\"word\": \"is\", \"dep_parent\": \"15\", \"pos\": \"VBZ\", \"char_idx\": \"93\", \"lemma\": \"be\", \"word_idx\": \"13\", \"dep_path\": \"auxpass\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"mutations\", \"dep_parent\": \"15\", \"pos\": \"NNS\", \"char_idx\": \"106\", \"lemma\": \"mutation\", \"word_idx\": \"16\", \"dep_path\": \"agent\", \"ner\": \"O\"}, \"children\": [{\"attrib\": {\"word\": \"PKD1\", \"dep_parent\": \"17\", \"pos\": \"NN\", \"char_idx\": \"123\", \"lemma\": \"pkd1\", \"word_idx\": \"19\", \"dep_path\": \"prep_in\", \"ner\": \"O\"}, \"children\": [{\"attrib\": {\"word\": \"the\", \"dep_parent\": \"20\", \"pos\": \"DT\", \"char_idx\": \"119\", \"lemma\": \"the\", \"word_idx\": \"18\", \"dep_path\": \"dep\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"PKD2\", \"dep_parent\": \"20\", \"pos\": \"NN\", \"char_idx\": \"131\", \"lemma\": \"pkd2\", \"word_idx\": \"21\", \"dep_path\": \"conj_or\", \"ner\": \"O\"}, \"children\": []}, {\"attrib\": {\"word\": \"genes\", \"dep_parent\": \"20\", \"pos\": \"NNS\", \"char_idx\": \"136\", \"lemma\": \"gene\", \"word_idx\": \"22\", \"dep_path\": \"dep\", \"ner\": \"O\"}, \"children\": []}]}]}]}]};\n",
       "\n",
       "// Constants\n",
       "var margin = {top: 20, right: 20, bottom: 20, left: 20},\n",
       "width = 800 - margin.left - margin.right,\n",
       "height = 250 - margin.top - margin.bottom,\n",
       "R = 5;\n",
       "\n",
       "// Create the d3 tree object\n",
       "var tree = d3.layout.tree()\n",
       "  .size([width, height]);\n",
       "\n",
       "// Create the svg canvas\n",
       "var svg = d3.select(\"#tree-chart-6391770092561218128\")\n",
       "  .append(\"svg\")\n",
       "  .attr(\"width\", width + margin.left + margin.right)\n",
       "  .attr(\"height\", height + margin.top + margin.bottom)\n",
       "  .append(\"g\")\n",
       "  .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
       "\n",
       "function renderTree() {\n",
       "  var nodes = tree.nodes(root),\n",
       "  edges = tree.links(nodes);\n",
       "\n",
       "  // Place the nodes\n",
       "  var nodeGroups = svg.selectAll(\"g.node\")\n",
       "    .data(nodes)\n",
       "    .enter().append(\"g\")\n",
       "    .attr(\"class\", \"node\")\n",
       "    .attr(\"transform\", function(d) { return \"translate(\" + d.x + \",\" + d.y + \")\"; });\n",
       "       \n",
       "  // Append circles\n",
       "  nodeGroups.append(\"circle\")\n",
       "    .on(\"click\", function() {\n",
       "      d3.select(this).classed(\"highlight\", !d3.select(this).classed(\"highlight\")); })\n",
       "    .attr(\"r\", R);\n",
       "     \n",
       "  // Append the actual word\n",
       "  nodeGroups.append(\"text\")\n",
       "    .text(function(d) { return d.attrib.word; })\n",
       "    .attr(\"text-anchor\", function(d) { \n",
       "      return d.children && d.children.length > 0 ? \"start\" : \"middle\"; })\n",
       "    .attr(\"dx\", function(d) { \n",
       "      return d.children && d.children.length > 0 ? R + 3 : 0; })\n",
       "    .attr(\"dy\", function(d) { \n",
       "      return d.children && d.children.length > 0 ? 0 : 3*R + 3; });\n",
       "\n",
       "  // Place the edges\n",
       "  var edgePaths = svg.selectAll(\"path\")\n",
       "    .data(edges)\n",
       "    .enter().append(\"path\")\n",
       "    .attr(\"class\", \"edge\")\n",
       "    .on(\"click\", function() {\n",
       "      d3.select(this).classed(\"highlight\", !d3.select(this).classed(\"highlight\")); })\n",
       "    .attr(\"d\", d3.svg.diagonal());\n",
       "}\n",
       "\n",
       "renderTree();\n",
       "});\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "xt = corenlp_to_xmltree(rows[1], prune_root=True)\n",
    "xt.render_tree()\n",
    "root = xt.root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Ngrams:lemma:FILTER-BY(pos=VB):BETWEEN-MENTION-and-MENTION, xpath=\"//*[{0}][1]/ancestor-or-self::*[count(. | //*[{1}][1]/ancestor-or-self::*) = count(//*[{1}][1]/ancestor-or-self::*)][1]/descendant-or-self::*[(count(.//*[{0}]) = count(//*[{0}])) or (count(.//*[{1}]) = count(//*[{1}]))][starts-with(@pos, 'VB')]\">\n",
      "LEMMA:FILTER-BY(pos=VB):BETWEEN-MENTION-and-MENTION[cause]\n"
     ]
    }
   ],
   "source": [
    "p1 = [0,1,2,3,4]\n",
    "g1 = [19]\n",
    "g2 = [21]\n",
    "ft = Ngrams(Filter(Between(Mention(0), Mention(1)), 'pos', 'VB'), 'lemma', 1)\n",
    "print ft\n",
    "ft.print_apply(root, [p1, g2], 'word_idx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WORD:BETWEEN-MENTION-and-MENTION[disorder_caused_mutations_PKD1]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[disorder_cause_mutation_pkd1]\n",
      "POS:BETWEEN-MENTION-and-MENTION[NN_VBN_NNS_NN]\n",
      "NER:BETWEEN-MENTION-and-MENTION[O_O_O_O]\n",
      "WORD:BETWEEN-MENTION-and-MENTION[disorder_caused_mutations]\n",
      "WORD:BETWEEN-MENTION-and-MENTION[caused_mutations_PKD1]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[disorder_cause_mutation]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[cause_mutation_pkd1]\n",
      "POS:BETWEEN-MENTION-and-MENTION[NN_VBN_NNS]\n",
      "POS:BETWEEN-MENTION-and-MENTION[VBN_NNS_NN]\n",
      "NER:BETWEEN-MENTION-and-MENTION[O_O_O]\n",
      "NER:BETWEEN-MENTION-and-MENTION[O_O_O]\n",
      "LEMMA:FILTER-BY(pos=VB):BETWEEN-MENTION-and-MENTION[cause]\n",
      "WORD:RIGHT-OF-MENTION[is]\n",
      "WORD:RIGHT-OF-MENTION[is_the]\n",
      "WORD:RIGHT-OF-MENTION[is_the_common]\n",
      "LEMMA:RIGHT-OF-MENTION[be]\n",
      "LEMMA:RIGHT-OF-MENTION[be_the]\n",
      "LEMMA:RIGHT-OF-MENTION[be_the_common]\n",
      "POS:RIGHT-OF-MENTION[VBZ]\n",
      "POS:RIGHT-OF-MENTION[VBZ_DT]\n",
      "POS:RIGHT-OF-MENTION[VBZ_DT_JJ]\n",
      "NER:RIGHT-OF-MENTION[O]\n",
      "NER:RIGHT-OF-MENTION[O_O]\n",
      "NER:RIGHT-OF-MENTION[O_O_O]\n",
      "WORD:RIGHT-OF-MENTION[genes]\n",
      "WORD:LEFT-OF-MENTION[the]\n",
      "WORD:RIGHT-OF-MENTION[genes]+WORD:LEFT-OF-MENTION[the]\n",
      "LEMMA:RIGHT-OF-MENTION[gene]\n",
      "LEMMA:LEFT-OF-MENTION[the]\n",
      "LEMMA:RIGHT-OF-MENTION[gene]+LEMMA:LEFT-OF-MENTION[the]\n",
      "POS:RIGHT-OF-MENTION[NNS]\n",
      "POS:LEFT-OF-MENTION[DT]\n",
      "POS:RIGHT-OF-MENTION[NNS]+POS:LEFT-OF-MENTION[DT]\n",
      "NER:RIGHT-OF-MENTION[O]\n",
      "NER:LEFT-OF-MENTION[O]\n",
      "NER:RIGHT-OF-MENTION[O]+NER:LEFT-OF-MENTION[O]\n"
     ]
    }
   ],
   "source": [
    "for feat in get_relation_features(root, p1, g2):\n",
    "    print feat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div class=\"table-wrapper\">\n",
       "    <h3>Causal genomic relationships</h3>\n",
       "    <table>\n",
       "        <tr><th>Gene</th><th>Variant</th><th>Phenotype</th></tr>\n",
       "        <tr><td>ABC</td><td><i>AG34</i></td><td>Headaches during defecation</td></tr>\n",
       "        <tr><td>BDF</td><td><i>CT2</i></td><td>Defecation during headaches</td></tr>\n",
       "        <tr><td>XYG</td><td><i>AT456</i></td><td>Defecasomnia</td></tr>\n",
       "    </table>\n",
       "</div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Some wishful thinking...\n",
    "table_xml = \"\"\"\n",
    "<div class=\"table-wrapper\">\n",
    "    <h3>Causal genomic relationships</h3>\n",
    "    <table>\n",
    "        <tr><th>Gene</th><th>Variant</th><th>Phenotype</th></tr>\n",
    "        <tr><td>ABC</td><td><i>AG34</i></td><td>Headaches during defecation</td></tr>\n",
    "        <tr><td>BDF</td><td><i>CT2</i></td><td>Defecation during headaches</td></tr>\n",
    "        <tr><td>XYG</td><td><i>AT456</i></td><td>Defecasomnia</td></tr>\n",
    "    </table>\n",
    "</div>\n",
    "\"\"\"\n",
    "from IPython.core.display import display_html, HTML\n",
    "display_html(HTML(table_xml))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tt = TableTree(table_xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tt.to_xml_str()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tt.root"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
