{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TreeDLib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We define three classes of operators:\n",
    "* _NodeSets:_ $S : 2^T \\mapsto 2^T$\n",
    "* _Indicators:_ $I : 2^T \\mapsto \\{0,1\\}^F$\n",
    "* _Combinators:_ $C : \\{0,1\\}^F \\times \\{0,1\\}^F \\mapsto \\{0,1\\}^F$\n",
    "\n",
    "where $T$ is a given input tree, and $F$ is the dimension of the feature space."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple demo: Generating DDLib features\n",
    "\n",
    "As a first simple demo let's generate the features generated by [ddlib](http://deepdive.stanford.edu/doc/basics/gen_feats.html).\n",
    "\n",
    "**_Note there are some noticeable differences stemming from using a dep tree representation at base; but a simple linear representation of the sentence could be used too, as in DDLib_**\n",
    "\n",
    "First, let's load a few sample sentences and convert one of them to XML format for testing; we'll also load the feature templates library, and also tag some candidates (crudely for now) to play around with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".node {\n",
       "      cursor: pointer;\n",
       "}\n",
       "\n",
       ".node circle {\n",
       "      fill: #fff;\n",
       "          stroke: steelblue;\n",
       "              stroke-width: 3px;\n",
       "}\n",
       "\n",
       ".node text {\n",
       "      font: 12px sans-serif;\n",
       "}\n",
       "\n",
       ".edge {\n",
       "      fill: none;\n",
       "          stroke: #ccc;\n",
       "              stroke-width: 2px;\n",
       "}\n",
       "</style>\n",
       "<div id=\"tree-chart\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "$.getScript(\"http://d3js.org/d3.v3.min.js\", function () {\n",
       "var root = {\"attrib\": {\"word\": \"disorder\", \"dep_parent\": 0, \"pos\": \"NN\", \"lemma\": \"disorder\", \"word_idx\": 80, \"ner\": \"O\"}, \"id\": 12, \"children\": [{\"attrib\": {\"word\": \"disease\", \"dep_parent\": 12, \"dep_label\": \"nsubj\", \"pos\": \"NN\", \"lemma\": \"disease\", \"word_idx\": 37, \"ner\": \"O\"}, \"id\": 5, \"children\": [{\"attrib\": {\"word\": \"Autosomal\", \"dep_parent\": 5, \"dep_label\": \"amod\", \"pos\": \"JJ\", \"lemma\": \"autosomal\", \"word_idx\": 0, \"ner\": \"O\"}, \"id\": 1, \"children\": []}, {\"attrib\": {\"word\": \"dominant\", \"dep_parent\": 5, \"dep_label\": \"amod\", \"pos\": \"JJ\", \"lemma\": \"dominant\", \"word_idx\": 10, \"ner\": \"O\"}, \"id\": 2, \"children\": []}, {\"attrib\": {\"word\": \"polycystic\", \"dep_parent\": 5, \"dep_label\": \"amod\", \"pos\": \"JJ\", \"lemma\": \"polycystic\", \"word_idx\": 19, \"ner\": \"O\"}, \"id\": 3, \"children\": []}, {\"attrib\": {\"word\": \"kidney\", \"dep_parent\": 5, \"dep_label\": \"nn\", \"pos\": \"NN\", \"lemma\": \"kidney\", \"word_idx\": 30, \"ner\": \"O\"}, \"id\": 4, \"children\": []}]}, {\"attrib\": {\"word\": \"is\", \"dep_parent\": 12, \"dep_label\": \"cop\", \"pos\": \"VBZ\", \"lemma\": \"be\", \"word_idx\": 45, \"ner\": \"O\"}, \"id\": 6, \"children\": []}, {\"attrib\": {\"word\": \"the\", \"dep_parent\": 12, \"dep_label\": \"det\", \"pos\": \"DT\", \"lemma\": \"the\", \"word_idx\": 48, \"ner\": \"O\"}, \"id\": 7, \"children\": []}, {\"attrib\": {\"word\": \"common\", \"dep_parent\": 12, \"dep_label\": \"amod\", \"pos\": \"JJ\", \"lemma\": \"common\", \"word_idx\": 57, \"ner\": \"O\"}, \"id\": 9, \"children\": [{\"attrib\": {\"word\": \"most\", \"dep_parent\": 9, \"dep_label\": \"advmod\", \"pos\": \"RBS\", \"lemma\": \"most\", \"word_idx\": 52, \"ner\": \"O\"}, \"id\": 8, \"children\": []}]}, {\"attrib\": {\"word\": \"human\", \"dep_parent\": 12, \"dep_label\": \"amod\", \"pos\": \"JJ\", \"lemma\": \"human\", \"word_idx\": 64, \"ner\": \"O\"}, \"id\": 10, \"children\": []}, {\"attrib\": {\"word\": \"monogenic\", \"dep_parent\": 12, \"dep_label\": \"amod\", \"pos\": \"JJ\", \"lemma\": \"monogenic\", \"word_idx\": 70, \"ner\": \"O\"}, \"id\": 11, \"children\": []}, {\"attrib\": {\"word\": \"caused\", \"dep_parent\": 12, \"dep_label\": \"conj_and\", \"pos\": \"VBN\", \"lemma\": \"cause\", \"word_idx\": 96, \"ner\": \"O\"}, \"id\": 15, \"children\": [{\"attrib\": {\"word\": \"is\", \"dep_parent\": 15, \"dep_label\": \"auxpass\", \"pos\": \"VBZ\", \"lemma\": \"be\", \"word_idx\": 93, \"ner\": \"O\"}, \"id\": 14, \"children\": []}, {\"attrib\": {\"word\": \"mutations\", \"dep_parent\": 15, \"dep_label\": \"agent\", \"pos\": \"NNS\", \"lemma\": \"mutation\", \"word_idx\": 106, \"ner\": \"O\"}, \"id\": 17, \"children\": [{\"attrib\": {\"word\": \"PKD1\", \"dep_parent\": 17, \"dep_label\": \"prep_in\", \"pos\": \"NN\", \"lemma\": \"pkd1\", \"word_idx\": 123, \"ner\": \"O\"}, \"id\": 20, \"children\": [{\"attrib\": {\"word\": \"the\", \"dep_parent\": 20, \"dep_label\": \"dep\", \"pos\": \"DT\", \"lemma\": \"the\", \"word_idx\": 119, \"ner\": \"O\"}, \"id\": 19, \"children\": []}, {\"attrib\": {\"word\": \"PKD2\", \"dep_parent\": 20, \"dep_label\": \"conj_or\", \"pos\": \"NN\", \"lemma\": \"pkd2\", \"word_idx\": 131, \"ner\": \"O\"}, \"id\": 22, \"children\": []}, {\"attrib\": {\"word\": \"genes\", \"dep_parent\": 20, \"dep_label\": \"dep\", \"pos\": \"NNS\", \"lemma\": \"gene\", \"word_idx\": 136, \"ner\": \"O\"}, \"id\": 23, \"children\": []}]}]}]}]}\n",
       "// See http://bl.ocks.org/d3noob/8375092\n",
       "\n",
       "// Tree variable is prepended to this file in python...\n",
       "// root = ...\n",
       "\n",
       "// Constants\n",
       "var margin = {top: 20, right: 20, bottom: 20, left: 20},\n",
       "width = 800 - margin.left - margin.right,\n",
       "height = 250 - margin.top - margin.bottom,\n",
       "R = 5;\n",
       "\n",
       "// Create the d3 tree object\n",
       "var tree = d3.layout.tree()\n",
       "  .size([width, height]);\n",
       "\n",
       "// Create the svg canvas\n",
       "var svg = d3.select(\"#tree-chart\")\n",
       "  .append(\"svg\")\n",
       "  .attr(\"width\", width + margin.left + margin.right)\n",
       "  .attr(\"height\", height + margin.top + margin.bottom)\n",
       "  .append(\"g\")\n",
       "  .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
       "\n",
       "function renderTree() {\n",
       "  var nodes = tree.nodes(root),\n",
       "  edges = tree.links(nodes);\n",
       "\n",
       "  // Place the nodes\n",
       "  var nodeGroups = svg.selectAll(\"g.node\")\n",
       "    .data(nodes)\n",
       "    .enter().append(\"g\")\n",
       "    .attr(\"class\", \"node\")\n",
       "    .attr(\"transform\", function(d) { return \"translate(\" + d.x + \",\" + d.y + \")\"; });\n",
       "       \n",
       "  // Append circles\n",
       "  nodeGroups.append(\"circle\")\n",
       "    .attr(\"r\", R);\n",
       "     \n",
       "  // Append the actual word\n",
       "  nodeGroups.append(\"text\")\n",
       "    .text(function(d) { return d.attrib.word; })\n",
       "    .attr(\"text-anchor\", function(d) { \n",
       "      return d.children && d.children.length > 0 ? \"start\" : \"middle\"; })\n",
       "    .attr(\"dx\", function(d) { \n",
       "      return d.children && d.children.length > 0 ? R + 3 : 0; })\n",
       "    .attr(\"dy\", function(d) { \n",
       "      return d.children && d.children.length > 0 ? 0 : 3*R + 3; });\n",
       "\n",
       "  // Place the edges\n",
       "  var edgePaths = svg.selectAll(\"path\")\n",
       "    .data(edges)\n",
       "    .enter().append(\"path\")\n",
       "    .attr(\"class\", \"edge\")\n",
       "    .attr(\"d\", d3.svg.diagonal());\n",
       "}\n",
       "\n",
       "renderTree();\n",
       "});\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from util import load_sentences, tag_candidate\n",
    "from tree_structs import DepTree, TableTree\n",
    "from feature_templates import *\n",
    "from basic_features import *\n",
    "dts = map(DepTree, load_sentences('test/test1.parsed.tsv'))\n",
    "dt = dts[1]\n",
    "tag_candidate(dt.root, ['Autosomal', 'dominant', 'polycystic', 'kidney', 'disease'], 'P1')\n",
    "tag_candidate(dt.root, ['PKD1'], 'G1')\n",
    "tag_candidate(dt.root, ['PKD2'], 'G2')\n",
    "dt.render_tree()\n",
    "#for i in range(len(dts)):\n",
    "#    dts[i].render_tree()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# TODO:\n",
    "# - XML -> JSON / visualization; be able to init Tree from XML!\n",
    "# - Clean up this notebook / code!\n",
    "\n",
    "def new_root_cid(root, cid):\n",
    "    \n",
    "    # Get candidate mention as new root\n",
    "    # NOTE: This will already contain all of its descendants from the old tree\n",
    "    new_root = root.xpath(\"//*[@cid='%s'][1]\" % cid)[0]\n",
    "    \n",
    "    # Recursively add parents + all their children, *minus current one!*\n",
    "    new_root.append(root.xpath(\"//*[@cid='%s'][1]/..\" % cid)[0])\n",
    "    return new_root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "root.xpath(\"//*[@cid='P1'][1]/..\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "et.tostring(new_root_cid(root, 'P1')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<MENTION, xpath=\"//*[@cid='{0}']\">"
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Mention(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LEMMA:BETWEEN-MENTION-and-MENTION[disorder]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[disease]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[cause]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[mutation]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[disorder_disease]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[disease_cause]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[cause_mutation]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[disorder_disease_cause]\n",
      "LEMMA:BETWEEN-MENTION-and-MENTION[disease_cause_mutation]\n"
     ]
    }
   ],
   "source": [
    "NGrams(Between(Mention(0), Mention(1)), 'lemma', 3).print_apply(root, ['P1', 'G1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Autosomal dominant polycystic kidney disease is the most common human monogenic disorder and is caused by mutations in the PKD1 or PKD2 genes.'"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "load_sentences('test/test1.parsed.tsv')[1].text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def flat_tree(root):\n",
    "    if root.get('dep_label') is not None:\n",
    "        s = '[%s]> %s' % (root.get('dep_label'), root.get('word'))\n",
    "    else:\n",
    "        s = root.get('word')\n",
    "    if len(root) > 0:\n",
    "        s += ' ( %s )' % ', '.join(flat_tree(c) for c in root)\n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'disorder ( [nsubj]> disease ( [amod]> Autosomal, [amod]> dominant, [amod]> polycystic, [nn]> kidney ), [cop]> is, [det]> the, [amod]> common ( [advmod]> most ), [amod]> human, [amod]> monogenic, [conj_and]> caused ( [auxpass]> is, [agent]> mutations ( [prep_in]> PKD1 ( [dep]> the, [conj_or]> PKD2, [dep]> genes ) ) ) )'"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ft = flat_tree(dt.root)\n",
    "ft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['disorder (  -> disease (  -> Autosomal -> dominant -> polycystic -> kidney )  -> is -> the -> common (  -> most )  -> human -> monogenic -> caused (  -> is -> mutations']"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.findall(r'disorder.*?caused.*?mutations', ft)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "xml = dt.to_xml_str()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<node dep_parent=\"0\" id=\"12\" lemma=\"disorder\" ner=\"O\" pos=\"NN\" word=\"disorder\" word_idx=\"80\"><node dep_label=\"nsubj\" dep_parent=\"12\" id=\"5\" lemma=\"disease\" ner=\"O\" pos=\"NN\" word=\"disease\" word_idx=\"37\" cid=\"P1\"><node dep_label=\"amod\" dep_parent=\"5\" id=\"1\" lemma=\"autosomal\" ner=\"O\" pos=\"JJ\" word=\"Autosomal\" word_idx=\"0\" cid=\"P1\"/><node dep_label=\"amod\" dep_parent=\"5\" id=\"2\" lemma=\"dominant\" ner=\"O\" pos=\"JJ\" word=\"dominant\" word_idx=\"10\" cid=\"P1\"/><node dep_label=\"amod\" dep_parent=\"5\" id=\"3\" lemma=\"polycystic\" ner=\"O\" pos=\"JJ\" word=\"polycystic\" word_idx=\"19\" cid=\"P1\"/><node dep_label=\"nn\" dep_parent=\"5\" id=\"4\" lemma=\"kidney\" ner=\"O\" pos=\"NN\" word=\"kidney\" word_idx=\"30\" cid=\"P1\"/></node><node dep_label=\"cop\" dep_parent=\"12\" id=\"6\" lemma=\"be\" ner=\"O\" pos=\"VBZ\" word=\"is\" word_idx=\"45\"/><node dep_label=\"det\" dep_parent=\"12\" id=\"7\" lemma=\"the\" ner=\"O\" pos=\"DT\" word=\"the\" word_idx=\"48\"/><node dep_label=\"amod\" dep_parent=\"12\" id=\"9\" lemma=\"common\" ner=\"O\" pos=\"JJ\" word=\"common\" word_idx=\"57\"><node dep_label=\"advmod\" dep_parent=\"9\" id=\"8\" lemma=\"most\" ner=\"O\" pos=\"RBS\" word=\"most\" word_idx=\"52\"/></node><node dep_label=\"amod\" dep_parent=\"12\" id=\"10\" lemma=\"human\" ner=\"O\" pos=\"JJ\" word=\"human\" word_idx=\"64\"/><node dep_label=\"amod\" dep_parent=\"12\" id=\"11\" lemma=\"monogenic\" ner=\"O\" pos=\"JJ\" word=\"monogenic\" word_idx=\"70\"/><node dep_label=\"conj_and\" dep_parent=\"12\" id=\"15\" lemma=\"cause\" ner=\"O\" pos=\"VBN\" word=\"caused\" word_idx=\"96\"><node dep_label=\"auxpass\" dep_parent=\"15\" id=\"14\" lemma=\"be\" ner=\"O\" pos=\"VBZ\" word=\"is\" word_idx=\"93\"/><node dep_label=\"agent\" dep_parent=\"15\" id=\"17\" lemma=\"mutation\" ner=\"O\" pos=\"NNS\" word=\"mutations\" word_idx=\"106\"><node dep_label=\"prep_in\" dep_parent=\"17\" id=\"20\" lemma=\"pkd1\" ner=\"O\" pos=\"NN\" word=\"PKD1\" word_idx=\"123\" cid=\"G1\"><node dep_label=\"dep\" dep_parent=\"20\" id=\"19\" lemma=\"the\" ner=\"O\" pos=\"DT\" word=\"the\" word_idx=\"119\"/><node dep_label=\"conj_or\" dep_parent=\"20\" id=\"22\" lemma=\"pkd2\" ner=\"O\" pos=\"NN\" word=\"PKD2\" word_idx=\"131\" cid=\"G2\"/><node dep_label=\"dep\" dep_parent=\"20\" id=\"23\" lemma=\"gene\" ner=\"O\" pos=\"NNS\" word=\"genes\" word_idx=\"136\"/></node></node></node></node>'"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<node dep_label=\"conj_or\" dep_parent=\"20\" id=\"22\" lemma=\"pkd2\" ner=\"O\" pos=\"NN\" word=\"PKD2\" word_idx=\"131\" cid=\"G2\"/>']"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.findall(r'<node[^>]*cid=\"G2\"[^>]*/>', xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def node(attribs):\n",
    "    return r'<[^>]+' + '\\s[^>]*\\s'.join('%s=\"%s\"' % (k,v) for k,v in attribs.iteritems()) + '[^>]*>'\n",
    "\n",
    "def child_of(attribs):\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<_sre.SRE_Match at 0x10ba0b718>"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.search(node({'cid':'G2'}), xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Element node at 0x10ba224b0>"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dt.root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LEMMA:BETWEEN-MENTION-and-MENTION[disorder_disease_cause_mutation]\n"
     ]
    }
   ],
   "source": [
    "Indicator(Between(Mention(0), Mention(1)), 'lemma').print_apply(dt.root, ['P1', 'G1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WORD:MENTION[PKD2]\n",
      "LEMMA:MENTION[pkd2]\n",
      "POS:MENTION[NN]\n",
      "NER:MENTION[O]\n",
      "WORD:MENTION[RGX:STARTS_W_CAPITAL=True]\n",
      "WORD:RIGHT-OF-MENTION[genes]\n",
      "WORD:LEFT-OF-MENTION[the]\n",
      "WORD:RIGHT-OF-MENTION[genes]+WORD:LEFT-OF-MENTION[the]\n",
      "LEMMA:RIGHT-OF-MENTION[gene]\n",
      "LEMMA:LEFT-OF-MENTION[the]\n",
      "LEMMA:RIGHT-OF-MENTION[gene]+LEMMA:LEFT-OF-MENTION[the]\n",
      "POS:RIGHT-OF-MENTION[NNS]\n",
      "POS:LEFT-OF-MENTION[DT]\n",
      "POS:RIGHT-OF-MENTION[NNS]+POS:LEFT-OF-MENTION[DT]\n",
      "NER:RIGHT-OF-MENTION[O]\n",
      "NER:LEFT-OF-MENTION[O]\n",
      "NER:RIGHT-OF-MENTION[O]+NER:LEFT-OF-MENTION[O]\n",
      "WORD:KEYWORD[monogenic]\n",
      "DEP_LABEL:BETWEEN-MENTION-and-KEYWORD[conj_and_agent_prep_in]\n",
      "LEMMA:BETWEEN-MENTION-and-KEYWORD[disorder_cause_mutation_pkd1]\n",
      "DEP_LABEL|LEMMA:BETWEEN-MENTION-and-KEYWORD[None|disorder_conj_and|cause_agent|mutation_prep_in|pkd1]\n"
     ]
    }
   ],
   "source": [
    "for feat in get_generic_mention_features(dt.root, 'G2', ['monogenic']):\n",
    "    print feat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div class=\"table-wrapper\">\n",
       "    <h3>Causal genomic relationships</h3>\n",
       "    <table>\n",
       "        <tr><th>Gene</th><th>Variant</th><th>Phenotype</th></tr>\n",
       "        <tr><td>ABC</td><td><i>AG34</i></td><td>Headaches during defecation</td></tr>\n",
       "        <tr><td>BDF</td><td><i>CT2</i></td><td>Defecation during headaches</td></tr>\n",
       "        <tr><td>XYG</td><td><i>AT456</i></td><td>Defecasomnia</td></tr>\n",
       "    </table>\n",
       "</div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Some wishful thinking...\n",
    "table_xml = \"\"\"\n",
    "<div class=\"table-wrapper\">\n",
    "    <h3>Causal genomic relationships</h3>\n",
    "    <table>\n",
    "        <tr><th>Gene</th><th>Variant</th><th>Phenotype</th></tr>\n",
    "        <tr><td>ABC</td><td><i>AG34</i></td><td>Headaches during defecation</td></tr>\n",
    "        <tr><td>BDF</td><td><i>CT2</i></td><td>Defecation during headaches</td></tr>\n",
    "        <tr><td>XYG</td><td><i>AT456</i></td><td>Defecasomnia</td></tr>\n",
    "    </table>\n",
    "</div>\n",
    "\"\"\"\n",
    "from IPython.core.display import display_html, HTML\n",
    "display_html(HTML(table_xml))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Causal genomic relationships\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-232-0fa4d2e68a78>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTableTree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtable_xml\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/aratner/repos/treedlib/tree_structs.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, raw)\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraw\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mraw\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_js\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mroot\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0mto_xml_str\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/aratner/repos/treedlib/tree_structs.py\u001b[0m in \u001b[0;36m_make_xml\u001b[0;34m(self, node)\u001b[0m\n\u001b[1;32m    104\u001b[0m     \u001b[0;31m# Recursively append children\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 106\u001b[0;31m       \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    107\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/aratner/repos/treedlib/tree_structs.py\u001b[0m in \u001b[0;36m_make_xml\u001b[0;34m(self, node)\u001b[0m\n\u001b[1;32m    104\u001b[0m     \u001b[0;31m# Recursively append children\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 106\u001b[0;31m       \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    107\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/aratner/repos/treedlib/tree_structs.py\u001b[0m in \u001b[0;36m_make_xml\u001b[0;34m(self, node)\u001b[0m\n\u001b[1;32m    103\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    104\u001b[0m     \u001b[0;31m# Recursively append children\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 105\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    106\u001b[0m       \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    107\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "tt = TableTree(table_xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<div class=\"table-wrapper\"><h3>Causal genomic relationships</h3><table><tr><th>Gene</th><th>Variant</th><th>Phenotype</th></tr><tr><td>ABC</td><td><i>AG34</i></td><td>Headaches during defecation</td></tr><tr><td>BDF</td><td><i>CT2</i></td><td>Defecation during headaches</td></tr><tr><td>XYG</td><td><i>AT456</i></td><td>Defecasomnia</td></tr></table></div>'"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt.to_xml_str()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Element div at 0x10ba22fa0>"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt.root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
